collectors/kvg/main.go

```diff
@@ -67,9 +77,69 @@ func main() {
 		return
 	}
 
+	// TODO: move to Subscriptions Handler package
+	type consumerEvent struct {
+		Stream   string `json:"stream"`
+		Consumer string `json:"consumer"`
+	}
+
+	consumers = make(map[string]string)
+	subjects = make(map[string]int)
+	subscriptions = []string{}
+
+	// already existing consumers
+	for consumerInfo := range c.JS.ConsumersInfo("data") {
+		consumers[consumerInfo.Name] = consumerInfo.Config.FilterSubject
+		subjects[consumerInfo.Config.FilterSubject]++
+		subscriptions = []string{}
+		for subject := range subjects {
+			subscriptions = append(subscriptions, subject)
+		}
+	}
+
+	// new consumers
+	c.Subscribe("$JS.EVENT.ADVISORY.CONSUMER.CREATED.>", func(msg *client.SubjectMessage) {
+		var consumerEvent consumerEvent
+		if err := json.Unmarshal([]byte(msg.Data), &consumerEvent); err != nil {
+			log.Fatalf("Parse response failed, reason: %v \n", err)
+		}
+		consumerInfo, _ := c.JS.ConsumerInfo(consumerEvent.Stream, consumerEvent.Consumer)
+		consumers[consumerInfo.Name] = consumerInfo.Config.FilterSubject
+		subjects[consumerInfo.Config.FilterSubject]++
+		subscriptions = []string{}
+		for subject := range subjects {
+			subscriptions = append(subscriptions, subject)
+		}
+		log.Debugln("Subscriptions", consumers)
+		log.Debugln("Subscriptions", subjects)
+		log.Debugln("Subscriptions", subscriptions)
+		collectors["map-stops"].Run()
+	})
+
+	// remove consumers
+	c.Subscribe("$JS.EVENT.ADVISORY.CONSUMER.DELETED.>", func(msg *client.SubjectMessage) {
+		var consumerEvent consumerEvent
+		if err := json.Unmarshal([]byte(msg.Data), &consumerEvent); err != nil {
+			log.Fatalf("Parse response failed, reason: %v \n", err)
+		}
+		if subjects[consumers[consumerEvent.Consumer]] > 1 {
+			subjects[consumers[consumerEvent.Consumer]]--
+		} else {
+			delete(subjects, consumers[consumerEvent.Consumer])
+		}
+		delete(consumers, consumerEvent.Consumer)
+		subscriptions = []string{}
```

Comment:
@lukashass Please add a `sync.Mutex` for `subscriptions`, `subjects` and `consumers` and lock if you change data.

Reply:
@anbraten What are the reasons for locking here?

One good reason I see is concurrent access by a callback writing into `subscriptions` and our scheduled collectors reading from `subscriptions`.

Another case I'm unsure about is concurrent access by two callbacks. nats is already doing some locking inside the subscriptions: https://github.com/nats-io/nats.go/blob/84e8b4dda7cca442de11abab3615665e7e3e258a/nats.go#L2615

Reply:
Added sync.Mutex in 26ea85f